internal/moduleref/resolver.go

```diff
@@ -44,40 +45,55 @@ func NewResolver(internalManifest modsdir.Manifest) *Resolver {
 // and return a new manifest encapsulating this information.
 func (r *Resolver) Resolve(cfg *configs.Config) *Manifest {
 	// First find all the referenced modules.
-	r.findAndTrimReferencedEntries(cfg)
+	r.findAndTrimReferencedEntries(cfg, nil)
 
 	return r.manifest
 }
 
 // findAndTrimReferencedEntries will traverse a given Terraform configuration
 // and attempt find a caller for every entry in the internal module manifest.
 // If an entry is found, it will be removed from the internal manifest and
-// appended to the manifest that records this new information.
-func (r *Resolver) findAndTrimReferencedEntries(cfg *configs.Config) {
-	for entryKey, entry := range r.internalManifest {
-		for callerKey := range cfg.Module.ModuleCalls {
-			// Construct the module path with the caller key to get
-			// the full module entry key. If it's a root module caller
-			// do nothing since the path will be empty.
-			path := strings.Join(cfg.Path, ".")
-			if path != "" {
-				callerKey = path + "." + callerKey
+// appended to the manifest that records this new information in a nested heirarchy.
+func (r *Resolver) findAndTrimReferencedEntries(cfg *configs.Config, parentRecord *Record) {
+	var name string
+	var versionConstraints version.Constraints
+	if cfg.Parent != nil {
+		for key, config := range cfg.Parent.Children {
+			if config.SourceAddr.String() == cfg.SourceAddr.String() {
```

Comment:
Asserting equal source addresses is problematic since there can be multiple declarations using the same source address. What will be unique per module declaration is the module's key. 

For example, with the following config:
```hcl
module "pets-can-code" {
  source  = "localterraform.com/hashicorp/pets-can-code/random"
  version = "0.1.0"
}

module "pets-can-code-2" {
  source  = "localterraform.com/hashicorp/pets-can-code/random"
  version = "0.1.0"
}
```

And I run `terraform modules`, returns:
```
.
├── "pets-can-code"[localterraform.com/hashicorp/pets-can-code/random] 0.1.0
├── "pets-can-code"[localterraform.com/hashicorp/pets-can-code/random] 0.1.0
```

Reply:
Nice catch, silly oversight in retrospect 

Included 👍 https://github.com/hashicorp/terraform/pull/36062/commits/40d1628e3c5b6d82b0c00f548cc76950ecf043b0